enterprise/server/remote_execution/execution_server/execution_server_test.go

```diff
@@ -0,0 +1,114 @@
+package execution_server
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/go-redis/redis/v8"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"gorm.io/gorm"
+
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/scheduling/scheduler_server"
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/scheduling/task_router"
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/testutil/testredis"
+	"github.com/buildbuddy-io/buildbuddy/enterprise/server/util/redisutil"
+	"github.com/buildbuddy-io/buildbuddy/server/tables"
+	"github.com/buildbuddy-io/buildbuddy/server/testutil/testenv"
+
+	"github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+)
+
+func executionServer(t *testing.T) *ExecutionServer {
+	env := testenv.GetTestEnv(t)
+
+	redisTarget := testredis.Start(t).Target
+	rdb := redis.NewClient(redisutil.TargetToOptions(redisTarget))
+	env.SetRemoteExecutionRedisClient(rdb)
+	env.SetRemoteExecutionRedisPubSubClient(rdb)
+
+	router, err := task_router.New(env)
+	if err != nil {
+		t.Fatal(err)
+	}
+	env.SetTaskRouter(router)
+
+	scheduler, err := scheduler_server.NewSchedulerServerWithOptions(env, &scheduler_server.Options{
+		EnableRedisAvailabilityMonitoring: true,
+		LocalPortOverride:                 80,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	env.SetSchedulerService(scheduler)
+
+	s, err := NewExecutionServer(env)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return s
+}
+
+func createExecution(t *testing.T, db *gorm.DB, execution *tables.Execution) {
+	err := db.Create(execution).Error
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func createExecutionRedis(t *testing.T, redisClient redis.UniversalClient, executionID string) {
```

Comment:
I don't love the idea of manually crafting synthetic data in redis in the test. We have a slightly higher level test in remote_execution_test that tests cancellation (TestInvocationCancellation). I wonder if it doesn't trigger the issue because that test only creates a single execution. What do you think of modifying that test to create multiple executions?

Reply:
Is the concern that the unit test shouldn't actually be writing to the db?

Can (and will) definitely modify that test too. Verified that adding a second execution would've caused that test to timeout. 

Personally I like having another unit test for execution_server::Cancel. If it fails, it becomes really clear where the bug is. Even if remote_execution_test::TestInvocationCancellation had failed, there's more ambiguity whether the bug is how the server handles Interrupted BES events for example. But can look into mocking out redis if that addresses the concern

Reply:
Writing to the DB is fine, but I don't like that the unit test is crafting data (redis structure) that is normally an internal detail of the scheduler service. One way to avoid this is to replace the real scheduler implementation with a fake in your test so that the test is only testing the behavior of execution_server and not of scheduler_server.

Reply:
I see - good call. Done